src/linear_algebra/lagrange.lean

```diff
@@ -1,224 +1,567 @@
+  (hst : s ⊆ t) : interpolate t v r =
+  ∑ i in s, (interpolate (insert i (t \ s)) v r) * basis s v i :=
+begin
+  symmetry,
+  refine eq_interpolate_of_eval_eq _ hvt (lt_of_le_of_lt (degree_sum_le _ _) _) (λ i hi, _),
+  { simp_rw [(finset.sup_lt_iff (with_bot.bot_lt_coe t.card)), degree_mul],
+    intros i hi,
+    have hs : 1 ≤ s.card := nonempty.card_pos ⟨_, hi⟩,
+    have hst' : s.card ≤ t.card := card_le_of_subset hst,
+    have H : t.card = (1 + (t.card - s.card)) + (s.card - 1),
+    { rw [add_assoc, tsub_add_tsub_cancel hst' hs, ← add_tsub_assoc_of_le (hs.trans hst'),
+          nat.succ_add_sub_one, zero_add] },
+    rw [degree_basis (set.inj_on.mono hst hvt) hi, H, with_bot.coe_add,
+        with_bot.add_lt_add_iff_right (@with_bot.coe_ne_bot _ (s.card - 1))],
+    convert degree_interpolate_lt _ (hvt.mono (coe_subset.mpr (insert_subset.mpr
+      ⟨hst hi, sdiff_subset _ _⟩))),
+    rw [card_insert_of_not_mem (not_mem_sdiff_of_mem_right hi), card_sdiff hst, add_comm] },
+
+  { simp_rw [eval_finset_sum, eval_mul],
+    by_cases hi' : i ∈ s,
+    { rw [← add_sum_erase _ _ hi', eval_basis_eq (hvt.mono hst) hi',
+          eval_interpolate_at_node _ (hvt.mono (coe_subset.mpr
+            (insert_subset.mpr ⟨hi, sdiff_subset _ _⟩))) (mem_insert_self _ _),
+          mul_one, add_right_eq_self],
+      refine sum_eq_zero (λ j hj, _),
+      rcases mem_erase.mp hj with ⟨hij, hj⟩,
+      rw [eval_basis_ne hij hi', mul_zero] },
+    { have H : ∑ j in s, eval (v i) (basis s v j) = 1,
+      { rw [← eval_finset_sum, sum_basis (hvt.mono hst) hs, eval_one] },
+      rw [← mul_one (r i), ← H, mul_sum],
+      refine sum_congr rfl (λ j hj, _),
+      congr,
+      exact eval_interpolate_at_node _ (hvt.mono (insert_subset.mpr ⟨hst hj, sdiff_subset _ _⟩))
+                                (mem_insert.mpr (or.inr (mem_sdiff.mpr ⟨hi, hi'⟩))) } }
+end
+
+theorem interpolate_eq_add_interpolate_erase (hvs : set.inj_on v s) (hi : i ∈ s) (hj : j ∈ s)
+  (hij : i ≠ j) : interpolate s v r = interpolate (s.erase j) v r * basis_divisor (v i) (v j) +
+  interpolate (s.erase i) v r * basis_divisor (v j) (v i) :=
+begin
+  rw [interpolate_eq_sum_interpolate_insert_sdiff _ hvs ⟨i, (mem_insert_self i {j})⟩ _,
+      sum_insert (not_mem_singleton.mpr hij), sum_singleton,
+      basis_doubleton_left hij, basis_doubleton_right hij,
+      finset.sdiff_insert_insert_of_mem_of_not_mem hi (not_mem_singleton.mpr hij),
+      sdiff_singleton_eq_erase, finset.doubleton_comm,
+      finset.sdiff_insert_insert_of_mem_of_not_mem hj (not_mem_singleton.mpr hij.symm),
+      sdiff_singleton_eq_erase],
+  { exact insert_subset.mpr ⟨hi, singleton_subset_iff.mpr hj⟩ },
+end
+
+end interpolate
+
+section nodal
+open finset polynomial
+variables {ι : Type*} {s : finset ι} {v : ι → F} {i : ι} (r : ι → F) {x : F}
+
+/--
+We define `nodal`, the unique monic polynomial whose roots are the nodes defined by `v` and `s`.
```

Comment:
I didn't immediately understand this use of "node", maybe add an explanation in plainer words?
```suggestion
`nodal s v` is the unique monic polynomial whose roots are the nodes defined by `v` and `s`.

That is, the roots of `nodal s v` are exactly the image of `v` on `s`, with appropriate multiplicity.
```